src/briefcase/integrations/base.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import annotations
+
+import os
+import platform
+import shutil
+import sys
+from abc import ABC
+from collections import defaultdict
+from collections.abc import Mapping
+from pathlib import Path
+from typing import DefaultDict
+
+import requests
+from cookiecutter.main import cookiecutter
+
+from briefcase.config import AppConfig
+from briefcase.console import Console, Log
+from briefcase.integrations.download import Download
+from briefcase.integrations.subprocess import Subprocess
+
+
+# TODO: Implement Tool base class
+class Tool(ABC):
+    """Tool Base."""
+
+
+class ToolCache(Mapping):
```

Comment:
Rationale for implementing as a `Mapping`:

The original design proposed multiple `ToolCache` objects for the commands to manage between "universal tools" and "app-bound tools". During implementation, it become clear that the two objects almost always needed to be managed together.

When a codeflow is using the app-bound tools, it will likely need access to the universal tools; that can only be done if both the  app-bound `ToolCache` and the universal `ToolCache` is passed around everywhere.

Therefore, as a simplification, I just combined the two concepts into one object where universal tools are accessed via attributes and app-bound tools are accessed via attributes on items of the `ToolCache`.

The implementation via `defaultdict` avoids the need to ensure the `ToolCache` is initialized for an individual app at arbitrary places throughout the Commands. This technically allows for infinite recursion but in practice that should not be done.

Considered but rejected:
I initially allowed something like `self.tools[app].subprocess` to fall back to `self.tools.subprocess` to allow more transparent fallbacks in the tools. However, this kept creating subtle bugs and unexpected behavior that it seemed more dangerous than useful.